File: src/_01_CreationalPattern/_01_03_AbstractFactory/WhiteshipFactory.js
import ShipFactory from "./ShipFactory.js";
import Whiteship from "./WhiteShip.js";

/**
 * @typedef { import("./shipPartsFactory").default } ShipPartsFactory
 */

export default class WhiteshipFactory extends ShipFactory {
  /** @type { ShipPartsFactory } */
  #shipPartsFactory;

  /**
   * @param { ShipPartsFactory } shipPartsFactory;
   */
  constructor(shipPartsFactory) {
    super();
    this.#shipPartsFactory = shipPartsFactory;
  }
  
  /**
   * @override
   * @param { string } name
   */
  _createShip(name) {
    const ship = new Whiteship(name);
    ship.setAnchor(this.#shipPartsFactory.createAnchor());
    ship.setWheel(this.#shipPartsFactory.createWheel());

    return ship;
  }
}

File: src/_01_CreationalPattern/_01_03_AbstractFactory/java/after/WhiteshipFactory.java
package _01_CreationalPattern._01_03_AbstractFactory.java.after;

public class WhiteshipFactory implements ShipFactory {

  private ShipPartsFactory shipPartsFactory;
  
  WhiteshipFactory(ShipPartsFactory shipPartsFactory) {
    this.shipPartsFactory = shipPartsFactory;
  }
  
  @Override
  public Ship createShip(String name) {
    Ship ship = new Whiteship(name);
    ship.setAnchor(shipPartsFactory.createAnchor());
    ship.setWheel(shipPartsFactory.createWheel());

    return ship;
  }
  
}
